File: ios/Classes/Models/Accounts/Account/AccountInformation.swift
// Copyright 2022 Pera Wallet, LDA

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//
//  AccountInformation.swift

import Foundation
import UIKit

typealias PublicKey = String
typealias RekeyDetail = [PublicKey: LedgerDetail]

final class AccountInformation: Codable {
    let address: String
    var name: String
    var ledgerDetail: LedgerDetail?
    var receivesNotification: Bool
    var rekeyDetail: RekeyDetail?
    var preferredOrder: Int
    var isBackedUp: Bool

    static let invalidOrder = -1

    var isWatchAccount: Bool {
        get {
            return type == .watch
        }
        set {
            type = newValue ? .watch : .standard
        }
    }

    private var type: AccountType
    
    init(
        address: String,
        name: String,
        isWatchAccount: Bool,
        ledgerDetail: LedgerDetail? = nil,
        rekeyDetail: RekeyDetail? = nil,
        receivesNotification: Bool = true,
        preferredOrder: Int? = nil,
        isBackedUp: Bool
    ) {
        self.address = address
        self.name = name
        self.type = isWatchAccount ? .watch : .standard
        self.ledgerDetail = ledgerDetail
        self.receivesNotification = receivesNotification
        self.rekeyDetail = rekeyDetail
        self.preferredOrder = preferredOrder ?? Self.invalidOrder
        self.isBackedUp = isBackedUp
    }
    
    required init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        address = try container.decode(String.self, forKey: .address)
        name = try container.decode(String.self, forKey: .name)
        type = try container.decodeIfPresent(AccountType.self, forKey: .type) ?? .standard
        ledgerDetail = try container.decodeIfPresent(LedgerDetail.self, forKey: .ledgerDetail)
        receivesNotification = try container.decodeIfPresent(Bool.self, forKey: .receivesNotification) ?? true
        rekeyDetail = try container.decodeIfPresent(RekeyDetail.self, forKey: .rekeyDetail)
        preferredOrder = try container.decodeIfPresent(Int.self, forKey: .preferredOrder) ?? Self.invalidOrder
        isBackedUp = try container.decodeIfPresent(Bool.self, forKey: .isBackedUp) ?? true
    }
}

extension AccountInformation {
    func updateName(_ name: String) {
        self.name = name
    }
    
    func addRekeyDetail(_ ledgerDetail: LedgerDetail, for address: String) {
        if rekeyDetail != nil {
            self.rekeyDetail?[address] = ledgerDetail
        } else {
            self.rekeyDetail = [address: ledgerDetail]
        }
    }

    func addRekeyDetail(_ rekeyDetail: RekeyDetail, for address: String) {
        if self.rekeyDetail != nil {
            self.rekeyDetail?[address] = rekeyDetail[address]
        } else {
            self.rekeyDetail = rekeyDetail
        }
    }
}

extension AccountInformation {
    enum AccountType:
        String,
        Codable {
        case standard = "standard"
        case watch = "watch"
        case ledger = "ledger"
        case rekeyed = "rekeyed"
    }
}

extension AccountInformation {
    enum CodingKeys: String, CodingKey {
        case address = "address"
        case name = "name"
        case type = "type"
        case ledgerDetail = "ledgerDetail"
        case receivesNotification = "receivesNotification"
        case rekeyDetail = "rekeyDetail"
        case preferredOrder = "preferredOrder"
        case isBackedUp = "isBackedUp"
    }
}

extension AccountInformation: Hashable {
    static func == (lhs: AccountInformation, rhs: AccountInformation) -> Bool {
        return lhs.address == rhs.address
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(address)
    }
}

File: android/app/src/main/java/com/algorand/android/models/AccountInformation.kt
/*
 * Copyright 2022 Pera Wallet, LDA
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 */

package com.algorand.android.models

import android.os.Parcelable
import com.algorand.android.models.AssetInformation.Companion.ALGO_ID
import com.algorand.android.models.Participation.Companion.DEFAULT_PARTICIPATION_KEY
import com.algorand.android.utils.AccountCacheManager
import com.algorand.android.utils.calculateMinBalance
import java.math.BigInteger
import java.math.BigInteger.ZERO
import kotlinx.parcelize.Parcelize

@Parcelize
data class AccountInformation(
    val address: String,
    val amount: BigInteger,
    val participation: Participation?,
    val rekeyAdminAddress: String?,
    private val allAssetHoldingMap: HashMap<Long, AssetHolding>,
    val createdAtRound: Long?,
    val appsLocalState: List<CreatedAppLocalState>? = null,
    val appsTotalSchema: CreatedAppStateScheme? = null,
    val appsTotalExtraPages: Int? = null,
    val totalCreatedApps: Int = 0,
    val lastFetchedRound: Long?
) : Parcelable {

    val assetHoldingMap: HashMap<Long, AssetHolding>
        get() = allAssetHoldingMap?.filterNot { it.value.isDeleted } as? HashMap<Long, AssetHolding> ?: hashMapOf()

    fun isCreated(): Boolean {
        return createdAtRound != null
    }

    fun setAssetHoldingStatus(assetId: Long, status: AssetStatus) {
        allAssetHoldingMap?.get(assetId)?.status = status
    }

    fun addPendingAssetHolding(assetHolding: AssetHolding) {
        if (!AssetStatus.isPending(assetHolding.status)) return
        allAssetHoldingMap?.put(assetHolding.assetId, assetHolding)
    }

    fun isRekeyed(): Boolean {
        return !rekeyAdminAddress.isNullOrEmpty() && rekeyAdminAddress != address
    }

    fun getAssetInformationList(accountCacheManager: AccountCacheManager): MutableList<AssetInformation> {
        val assetInformationList = mutableListOf<AssetInformation>()
        assetInformationList.add(
            AssetInformation.getAlgorandAsset(amount)
        )
        assetHoldingMap.values.forEach { assetHolding ->
            accountCacheManager.getAssetDescription(assetHolding.assetId)?.let { assetDescription ->
                assetInformationList.add(AssetInformation.createAssetInformation(assetHolding, assetDescription))
            }
        }
        return assetInformationList
    }

    fun getAllAssetIds(): List<Long> {
        return assetHoldingMap.keys.toList()
    }

    fun getAllAssetIdsIncludeAlgorand(): List<Long> {
        return getAllAssetIds().toMutableList().apply { add(0, ALGO_ID) }
    }

    fun getOptedInAssetsCount() = allAssetHoldingMap?.size ?: 0

    fun getMinAlgoBalance(): BigInteger {
        return calculateMinBalance(
            this,
            isRekeyed() || isThereAnyDifferentAsset() || isThereAnOptedInApp()
        ).toBigInteger()
    }

    fun isAssetSupported(assetId: Long): Boolean {
        return assetId == ALGO_ID || hasAsset(assetId)
    }

    fun getBalance(assetId: Long): BigInteger {
        return if (assetId == ALGO_ID) {
            amount
        } else {
            getAssetHoldingOrNull(assetId)?.amount ?: ZERO
        }
    }

    fun doesUserHasParticipationKey() =
        !(participation == null || participation.voteParticipationKey == DEFAULT_PARTICIPATION_KEY)

    fun isThereAnyDifferentAsset() = assetHoldingMap.isNotEmpty()

    fun isThereAnOptedInApp() = appsLocalState?.isNotEmpty() == true || totalCreatedApps > 0

    fun hasAsset(assetId: Long): Boolean {
        return assetHoldingMap.containsKey(assetId)
    }

    fun getAssetHoldingOrNull(assetId: Long): AssetHolding? {
        return assetHoldingMap.get(assetId)
    }

    fun getAssetStatusOrNull(assetId: Long): AssetStatus? {
        return getAssetHoldingOrNull(assetId)?.status
    }

    fun getAssetHoldingList(): List<AssetHolding> {
        return assetHoldingMap.values.toList()
    }

    fun getAssetIdList(): List<Long> {
        return assetHoldingMap.keys.toList()
    }
}
